### DISCUSSION TITLE
Observing MIDIManager Endpoints

### DISCUSSION THREAD

**Discussion by fracturedsoftware** [orchetect/swift-midi#179] [Q&A] [answered]
I am trying to independently observe changes to midiManager.endpoints.inputsUnowned and midiManager.endpoints.outputsUnowned

In a SwiftUI view, assuming the midiManager is in the environment, I can do:

```swift
.onChange(of: midiManager.endpoints.inputsUnowned) {
    print("Unowned Inputs: \($0.map{ $0.displayName })")
}
```

Which works fine. However I am having difficulty figuring out how to do this within a class.

Normally I have a function in my class a bit like:

```swift
func observeObject(property: Published<Object>.Publisher) {  
    property.sink { [self] in
        print($0)
    }
    .store(in: &subs)
}
```

And - assuming the property is `@Published` - call it like: 

```swift
observeObject(property: myClass.$object)
```

But I am unable to figure out how to do the same with midiManager.endpoints so that I can observe them:

```swift
observeUnownedInputs(property: midiManager.endpoints.inputsUnowned)

func observeUnownedInputs(property: Published<[MIDIInputEndpoint]>.Publisher) {
    property.sink{ [self] in
        print($0)
    }
    .store(in: &subs)
}
```

Any help with getting this working would be greatly appreciated.

Thanks!

**Comment by orchetect** [comment]
Likely because the internal properties of `midiManager.endpoints` are not `@Published`. This was to maintain backwards compatibility with older OS versions that do not support it.

I just tried a few things and can't seem to find a quick solution.

**Reply by orchetect** [reply]
In the meantime you could use the notification handler in `MIDIManager` to be notified when endpoints change and work your internal updates around that.

**Reply by fracturedsoftware** [reply]
Ok, thanks. 
It looks like the notification method would require a lot of parsing so I'm feeding the result of the onChange() event to my MIDIHelper class as above. Works well.

**Reply by orchetect** [reply]
Doesn't require any parsing at all if you're just interested in knowing when endpoints change. You can just fetch the endpoints whenever added/removed/setupChanged notifications come through and ignore the associated values.

**Reply by orchetect** [reply]
I did some testing and there's a way to get this to work with some modifications to the library.

On the [endpoints-publisher](https://github.com/orchetect/MIDIKit/tree/endpoints-publisher) branch I've added `@Published` to the array properties of `midiManager.endpoints`.

This is by way of `midiManager.endpoints` being of a protocol type, which was originally intentioned for this very scenario. It means that on systems that support Combine (> macOS 10.15, iOS 13) the `@Published` wrapper could be added conditionally internally.

The downside is that casting is required to allow the compiler to know that the published properties are being accessed.

This works:

```swift
if let endpoints = midiManager.endpoints as? MIDIPublishedEndpoints {
     endpoints.$inputsUnowned
        .sink { array in
            print(array)
        }
        .store(in: &subs)
} else {
    // solution for pre-macOS 10.15 / pre-iOS 13
}
```

**Reply by fracturedsoftware** [reply]
Hi Steffan,

Thanks for this. Unfortunately I am away on holidays until end of May so am unable to test. When I get back I’ll take a look and also get back to you re my use case for the endpoints in my other question.

Cheers

**Reply by orchetect** [reply, marked as answer]
I have a solution for this that will work more elegantly.

A new subclass of `MIDIManager` called `ObservableMIDIManager` can be used in its place, which adds two new `@Published` properties:
- `observableDevices`
- `observableEndpoints`

Which means it's as simple as:
- swapping `let midiManager = MIDIManager( ... )` init with `ObservableMIDIManager( ... )`
- accessing its `observableEndpoints` property

Still putting finishing touches on, but will be in the next release.
